include/cuda/memory_resource

```diff
@@ -0,0 +1,330 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+#include "stream_view"
+#include "std/version"
+#include "std/cstddef"
+#include "std/detail/__config"
+#include "std/detail/__pragma_push"
+#include "std/type_traits"
+#include "std/cstddef"
+
+#if _LIBCUDACXX_STD_VER > 14
+#include <memory_resource>
+#endif // _LIBCUDACXX_STD_VER > 14
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/**
+ * @brief Specifies the kind of memory of an allocation.
+ *
+ * Memory allocation kind determines where memory can be accessed and the
+ * performance characteristics of accesses.
+ *
+ */
+enum class memory_kind {
+  device,  ///< Device memory accessible only from device
+  unified, ///< Unified memory accessible from both host and device
+  pinned,  ///< Page-locked system memory accessible from both host and device
+  host     ///< System memory only accessible from host code
+};
+
+/**
+ * @brief Tag type for the default context of `memory_resource`.
+ *
+ * Default context in which storage may be used immediately on any thread or any
+ * CUDA stream without synchronization.
+ */
+struct any_context{};
+
+namespace detail {
+struct __empty {};
+
+template <typename _Context> 
+struct __get_context_impl {
+  _Context get_context() const noexcept { return do_get_context(); }
+private:
+  virtual _Context do_get_context() const noexcept = 0;
+};
+
+// Specialization for the default `any_context` returns a default constructed
+// object and removes the virtual `do_get_context` from the class
+template<>
+struct __get_context_impl<any_context> {
+  any_context get_context() const noexcept { return any_context{}; }
+};
+
+} // namespace detail
+
+/**
+ * @brief Abstract interface for context specific memory allocation.
+ *
+ * @tparam _Kind The `memory_kind` of the allocated memory.
+ * @tparam _Context The execution context on which the storage may be used
+ * without synchronization
+ */
+template <memory_kind _Kind, typename _Context = any_context>
+class memory_resource : public detail::__get_context_impl<_Context>
+#if _LIBCUDACXX_STD_VER > 14
+     , public _CUDA_VSTD::conditional_t<
+          (_Kind == memory_kind::host) and
+              (_CUDA_VSTD::is_same<_Context, any_context>::value),
+          std::pmr::memory_resource, detail::__empty>
```

Comment:
Issues:
1. Inheriting from pmr::memory_resource makes violates ODR when there are TUs/libraries compiled with different _LIBCUDAXCC_STD_VER
2. ABI actually changes when adding pmr::memory_resource because of pmr::memory_resource::do_is_equal, which is (obviously) not present when there's no pmr::memory_resource
3. ABI changes _drastically_ when adding the get_context_impl class _and_ when there's pmr::memory_resource base - there are 2 base classes with virtual function table and the layout of the object is different depending on the presence of pmr::memory_resource class.

Reply:
Removed inheritance from `std::pmr::memory_resource`.